lib/ansible/modules/cloud/amazon/iam_role.py

```diff
@@ -202,32 +208,34 @@ def create_or_update_role(connection, module):
     # Check attached managed policies
     current_attached_policies = get_attached_policy_list(connection, params['RoleName'])
     if not compare_attached_role_policies(current_attached_policies, managed_policies):
-        # If managed_policies has a single empty element we want to remove all attached policies
-        if len(managed_policies) == 1 and managed_policies[0] == "":
+        if module.params.get('remove_unlisted_policies'):
+            # If managed_policies has a single empty element we want to remove all attached policies
+            if len(managed_policies) == 1 and managed_policies[0] == "":
+                for policy in current_attached_policies:
+                    try:
+                        connection.detach_role_policy(RoleName=params['RoleName'], PolicyArn=policy['PolicyArn'])
+                    except (ClientError, ParamValidationError) as e:
+                        module.fail_json(msg=e.message, **camel_dict_to_snake_dict(e.response))
+
+            # Detach policies not present
```

Comment:
This is a preexisting comment, but yes, it does appear to describe the purpose of the next two for loops.